### PR TITLE
CXX-2753 Refactor CMake config files to reflect new directory structure

### PR DESCRIPTION
**Pull request by eramongodb**
## Summary

This PR is a followup to https://github.com/mongodb/mongo-cxx-driver/pull/1026 and is related to CXX-2753. Verified by [this patch](https://spruce.mongodb.com/version/65242f4cd1fe0793d184ff09).

This PR refactors the CMake config files to better reflect and take advantage of the new directory structure by moving variables and commands into the scope of per-project subdirectories as much as possible (`cmake`, `include`, `lib`, `test`). This PR also takes this opportunity to apply a consistent formatting to the modified CMake config files (whitespacing, indentation, etc.).

Due to the already large diff in this PR, the relocation of currently out-of-place test files (under the `test_util` directory, i.e. `catch.hh`) is deferred to a followup PR.

This PR does/should _**NOT**_ affect binary or source compatibility (verified by [this patch](https://spruce.mongodb.com/task/cxx_driver_abi_compliance_check_abi_compliance_check_patch_1670696738466769e8b43af5561afc10cfed77cc_65242f961e2d17ef82c4d29c_23_10_09_16_51_37/logs?execution=0)).

## Notable Changes and Patterns

- `if(1)` + `endif()` is used to "scope" related commands together for readability purposes.
- Fixed redundant/incorrect regeneration of some generated files. Commands for every file that should be generated _once_ is now only run _once_ (e.g. export header, package targets file, etc.).
- Globbing of source files has been replaced with explicit, sorted lists to abide by [CMake recommendations](https://cmake.org/cmake/help/v3.15/command/file.html#glob).
- pkg-config config files were updated to use `CMAKE_INSTALL_INCLUDEDIR` instead of assuming `/include`.
- Relocation of test targets into `test` subdirectory also relocates their output files, which required updating the EVG config to include the test subdirectories in library/executable lookup.
- `<name>_test_properties` targets are used to define and set common target properties for tests.
- All CMake `message()` commands without a mode (aka message type) have been given `STATUS` mode, as otherwise their output is written to stderr (as-if NOTICE mode) which seems undesirable (status quo).
- Replaced generator-specific parallelism flags with `CMAKE_BUILD_PARALLEL_LEVEL`.

### CODE REVIEW

**Review by kevinAlbs — APPROVED** [review]
The reorganization and improvements are appreciated. LGTM with minor comments.

What tool was used to format the cmake files?

Aside: I like the `if(1)` / `endif()` pattern to separate blocks. When the C++ driver minimum CMake is increased to 3.25 or higher, the [`block` command](https://cmake.org/cmake/help/latest/command/block.html) may help.

File: src/bsoncxx/test/CMakeLists.txt [hunk 1]
```diff
@@ -16,23 +16,55 @@ if(NOT BUILD_TESTING)
...
-target_include_directories(test_bson PRIVATE ${libbson_include_directories})
-target_compile_definitions(test_bson PRIVATE ${libbson_definitions})
+# Common target properties for test executables.
+add_library(bsoncxx_test_properties INTERFACE)
+target_sources(bsoncxx_test_properties INTERFACE ${THIRD_PARTY_SOURCE_DIR}/catch/main.cpp)
+target_link_libraries(bsoncxx_test_properties INTERFACE bsoncxx_testing ${libbson_target})
+target_include_directories(bsoncxx_test_properties INTERFACE ${libbson_include_directories})
+target_compile_definitions(bsoncxx_test_properties INTERFACE ${libbson_definitions})
+
+target_link_libraries(test_bson PRIVATE bsoncxx_test_properties)
```
[the remark this answers is not in the document]
Comment (eramongodb) [reply src/bsoncxx/test/CMakeLists.txt:52]: This was done to have symmetry with the mongocxx test config and to accomodate the potential addition of more bsoncxx test executables beyond `test_bson`. Do you still prefer its removal?

[the remark this answers is not in the document]
Comment (kevinAlbs) [reply src/bsoncxx/test/CMakeLists.txt:52]: No strong preference. `bsoncxx_test_properties` may be useful if more test targets are added. Resolving.

**Comment by eramongodb** [comment]
> What tool was used to format the cmake files?

I am using the [CMake Language Support](https://github.com/josetr/cmake-language-support) extension for [VS Code](https://marketplace.visualstudio.com/items?itemName=josetr.cmake-language-support-vscode).

Note: it supports a `cmake.format.spaceAfterCommandName` parameter if we want to preserve the space before opening parenthesis (e.g. `find_package (bson-1.0 REQUIRED)`). The changes in this PR use the default value of `false` (e.g. `find_package(bson-1.0 REQUIRED)`). There are a mix of both styles in the config files, so if the space is preferred, that can be applied instead.

**Review by vector-of-bool — APPROVED** [review]
Sorry to delay. The CMake formatting changes are +1 good.

**Pull request merged**